indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java

```diff
@@ -279,53 +279,65 @@ public static void addSegment(
       String subTaskId,
       DataSegment segment,
       File segmentFile
-  ) throws IOException
-  {
-    final StorageLocation location = findLocationForSegment(cyclicIterator, numLocations, segment);
-    final File destFile = new File(
-        getPartitionDir(location, supervisorTaskId, segment.getInterval(), segment.getShardSpec().getPartitionNum()),
-        subTaskId
-    );
-    FileUtils.forceMkdirParent(destFile);
-    final long copiedBytes = Files.asByteSource(segmentFile).copyTo(Files.asByteSink(destFile));
-    if (copiedBytes == 0) {
-      throw new IOE(
-          "0 bytes copied after copying a segment file from [%s] to [%s]",
-          segmentFile.getAbsolutePath(),
-          destFile.getAbsolutePath()
-      );
-    }
-    location.addFile(destFile);
-  }
-
-  private static StorageLocation findLocationForSegment(
-      Iterator<StorageLocation> cyclicIterator,
-      int numLocations,
-      DataSegment segment
   )
   {
     for (int i = 0; i < numLocations; i++) {
       final StorageLocation location = cyclicIterator.next();
-      if (location.canHandle(segment)) {
-        return location;
+      final File destFile = location.reserve(
+          getPartitionFilePath(
+              supervisorTaskId,
+              subTaskId,
+              segment.getInterval(),
+              segment.getShardSpec().getPartitionNum()
+          ),
+          segment.getId(),
+          segmentFile.length()
+      );
+      if (destFile != null) {
+        try {
+          FileUtils.forceMkdirParent(destFile);
+          final long copiedBytes = Files.asByteSource(segmentFile).copyTo(Files.asByteSink(destFile));
+          if (copiedBytes == 0) {
+            throw new IOE(
+                "0 bytes copied after copying a segment file from [%s] to [%s]",
+                segmentFile.getAbsolutePath(),
+                destFile.getAbsolutePath()
+            );
+          } else {
+            return;
+          }
+        }
+        catch (Exception e) {
```

Comment:
Can catch narrower exception type?

Reply:
This is to print something in the log file so that operators could notice that something happened.

Reply:
This is generally considered an antipattern to catch everything just in case unless the exception is rethrown in the catch block. Could catch only IOExceptions?

Reply:
Sorry, I was confused with what part you were saying. It sounds good. Fixed.